### PR TITLE
Contribution: package USER-HDNNP for high-dimensional neural network potentials

### PR DESCRIPTION
**Pull request by singraber**
**Summary**

I would like to propose the addition of a new user package "USER-NNP" which requires linkage to the external [n2p2](https://github.com/CompPhysVienna/n2p2) library. The package contains a new pair style "nnp" which allows to use high-dimensional neural network potentials ([Behler and Parrinello, Phys. Rev. Lett. 2007, 98 (14), 146401](https://doi.org/10.1103/PhysRevLett.98.146401)).  These machine learning potentials require training before they can be used in production MD runs. Training and related tasks can be performed with tools provided by the n2p2 package. However, n2p2 does not come with its own MD engine (why reinvent the wheel?) and instead relies on an interface with LAMMPS ([Singraber, Behler and Dellago, J. Chem. Theory Comput. 2019, 15 (3), 1827-1840](https://doi.org/10.1021/acs.jctc.8b00770)) . This has been shipped with n2p2 for quite a while now and I think it would be beneficial for users if it was finally integrated in LAMMPS directly. The n2p2 package is kept compatible with Jörg Behler's original HDNNP software [RuNNer](https://www.uni-goettingen.de/de/560580.html) regarding its predictive capabilities, i.e. a neural network potential trained with RuNNer can also use the same LAMMPS/n2p2 interface which further increases the potential user base. However, currently only short-range HDNNPs (the most common type used) are supported, but an extension to include electrostatics is work in progress. 

I have now integrated the USER-NNP package which was shipped previously with n2p2 directly in LAMMPS and tried to conform to coding standards and include all the required documentation. I hope I did not miss any major problems, however, there are a few open points (see below) for which I would like to ask for your opinion. Overall, I am looking forward to hear your suggestions and whether you think my proposal would fit into the LAMMPS infrastructure.

Thank you for your efforts!

**Related Issue(s)**

None

**Author(s)**

Andreas Singraber (andreas.singraber@gmx.at)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Not aware of any problems.

**Implementation Notes**

Compiling LAMMPS with the USER-NNP package requires also to build (parts of) n2p2 in advance. This is shortly described in the corresponding "Build Extra" section (3.7.23) of the manual. More detailed information can be found in the ``lib/nnp/README`` file. After compiling n2p2 the build process in LAMMPS should be straightforward, following the usual procedure to include user packages. There is one extra flag ``-D N2P2_DIR`` which must be set in case n2p2 is not located in ``lib/nnp/n2p2``. To make a serial LAMMPS build also n2p2 must be compiled with MPI disabled, see ``lib/nnp/README`` how to achieve that.

**IMPORTANT:** For now, please ignore the remarks about using n2p2 version ``v2.2.0`` and check out the ``master`` branch instead.

No other LAMMPS features are affected by this user package.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

As mentioned above I want to address a few issues which you may also consider:

- [x] **(1) Name of the package/pair style**

   Using NNP/nnp as package/pair style name sounds very general. If you are not happy with that this could be switched to HDNNP/hdnnp or some other descriptive abbreviation (considering that there is a similar project on the way: #2570).

- [x] **(2) CMake files**

   As far as I could test both the CMake and the traditional build method work (also with ``-DLAMMPS_BIGBIG``). However, I have never used CMake before, so there may be some weird/incorrect ways I handled the library search in ``cmake/Modules/FindN2P2.cmake`` and ``cmake/Modules/Packages/USER-NNP.cmake``.

- [x] **(3) Older LAMMPS-style code parts**

   Some parts in ``pair_nnp.cpp`` were written before the ``Tokenizer`` class existed and hence look a bit outdated (see ``PairNNP::settings()``). The code works as it is but if you prefer I can rewrite these parts.

- [x] **(4) Element/Type mapping**

   At some point I introduced the ``emap`` keyword which allows the user to manually assign n2p2 element strings to LAMMPS atom types. This is important because in n2p2 elements are always sorted according to atomic number while in LAMMPS an input configuration could define atoms in a different way. So I don't support any other ``pair_coeff`` call besides ``pair_coeff * * max_cutoff`` and instead use the user's ``emap`` input to sort out atom types I don't need (e.g. in a ``pair_style hybrid`` situation). I wonder if the ``setflag`` variable which is used in many pair styles is somehow related to this mapping problem? Would you prefer a solution which makes use of the ``setflag`` mechanism?

- [x] **(5) Pimpl idiom**

   When compiling ``pair_nnp.cpp`` the compiler needs to "see" a lot of n2p2 headers down to very basic ones. This is not a problem because they are automatically collected in n2p2's ``include`` directory when compiling ``libnnpif``. However, I think the "Pimpl idiom" would be a more clever solution here. Would you prefer/strictly require to implement the interface in this way?

### CODE REVIEW

**Review by akohlmey — CHANGES_REQUESTED** [review]
Thanks for you submission. I have not tried it out yet, but for the most part it looks pretty clean. I noticed a few cosmetic things and added comments in the sources. Will respond to your comments and questions in a separate message.

**Comment by akohlmey** [comment]
> As mentioned above I want to address a few issues which you may also consider:
> 
>  **(1) Name of the package/pair style**
>   Using NNP/nnp as package/pair style name sounds very general. If you are not happy with that this could be switched to HDNNP/hdnnp or some other descriptive abbreviation (considering that there is a similar project on the way: #2570).

Switching to `USER-HDNNP` and `pair style hdnnp` would be preferred. Due to the large number of research groups investing into this style of potentials there are several of those in development and a few submitted (and one temporarily withdrawn). So to make names less ambiguous is helpful. It should also help to make your contribution more recognizable!

> **(2) CMake files**
>   As far as I could test both the CMake and the traditional build method work (also with `-DLAMMPS_BIGBIG`). However, I have never used CMake before, so there may be some weird/incorrect ways I handled the library search in `cmake/Modules/FindN2P2.cmake` and `cmake/Modules/Packages/USER-NNP.cmake`.

This is something that @junghans and @rbberger can check out. They are usually pretty good at spotting problems.
One of the biggest concerns is usually maintaining compatibility with the range of Linux "flavors" that we want to be compatible to. And that currently sets the minimum CMake version to 3.10.

> **(3) Older LAMMPS-style code parts**
>   Some parts in `pair_nnp.cpp` were written before the `Tokenizer` class existed and hence look a bit outdated (see `PairNNP::settings()`). The code works as it is but if you prefer I can rewrite these parts.

This is not an urgent issue unless rewrites are required anyway. I can look at that later after the more general issues have been addressed and I have done some experiments and manual checks.

> **(4) Element/Type mapping**
>   At some point I introduced the `emap` keyword which allows the user to manually assign n2p2 element strings to LAMMPS atom types. This is important because in n2p2 elements are always sorted according to atomic number while in LAMMPS an input configuration could define atoms in a different way. So I don't support any other `pair_coeff` call besides `pair_coeff * * max_cutoff` and instead use the user's `emap` input to sort out atom types I don't need (e.g. in a `pair_style hybrid` situation). I wonder if the `setflag` variable which is used in many pair styles is somehow related to this mapping problem? Would you prefer a solution which makes use of the `setflag` mechanism?

See my comment embedded in the code. I am mostly thinking of how familiar LAMMPS users with specifying the element/label mapping on the `pair_coeff * * ` line, while the global cutoff would be specified with the pair_style command.

> **(5) Pimpl idiom**
>   When compiling `pair_nnp.cpp` the compiler needs to "see" a lot of n2p2 headers down to very basic ones. This is not a problem because they are automatically collected in n2p2's `include` directory when compiling `libnnpif`. However, I think the "Pimpl idiom" would be a more clever solution here. Would you prefer/strictly require to implement the interface in this way?

What you currently have is already sufficiently pimpl-ified for our needs. There is little concern about which headers are included in the pair*.cpp files, especially when the include statement follows all LAMMPS headers. The problems arise with including headers in the pair*.h file (or people putting a "using namespace XXX;" there). That can lead to clashes with the way we currently built the style name to creating a class instance map.

File: cmake/Modules/FindN2P2.cmake
Comment (junghans) [inline cmake/Modules/FindN2P2.cmake:15]:
```suggestion
find_path(N2P2_INCLUDE_DIR InterfaceLammps.h HINTS "${N2P2_DIR}/include")
```

Comment (junghans) [reply cmake/Modules/FindN2P2.cmake:15]: Same for the others below.

File: cmake/Modules/FindN2P2.cmake
Comment (junghans) [inline cmake/Modules/FindN2P2.cmake:46]: I would create an imported target `N2P2::n2p2` here.

File: cmake/Modules/FindN2P2.cmake
Comment (junghans) [inline cmake/Modules/FindN2P2.cmake:8]: Not sure can rely on a manually downloaded package here, I would prefer a `ExternalProject_Add()`, have a look at `LATTE.cmake` as an example.

**Comment by akohlmey** [comment]
@singraber this has been dormant for about 3 weeks. Can please give us a quick update on your plans?
Please note that we currently have multiple pull requests with neural network based pair styles pending and ideally we want to do the final review and processing of all of them for the same patch release so responses from our side may also be slow at times while we focus on ongoing refactoring work in the code base.

**Comment by singraber** [comment]
First of all, let me thank all of you who already gave valuable suggestions! I am sorry that I did not have the time to process this further for the last weeks. However, I am going to work on the changes starting this week.. since there are no major issues yet I expect that I will have it ready for another round of review until beginning of next week. @akohlmey Does this fit into the planning for the patch release? Is there an approximate date considered for this patch?

**Comment by akohlmey** [comment]
> First of all, let me thank all of you who already gave valuable suggestions! I am sorry that I did not have the time to process this further for the last weeks. However, I am going to work on the changes starting this week.. since there are no major issues yet I expect that I will have it ready for another round of review until beginning of next week.

No worries. Because of using an external library you PR is ahead of the curve, since there are fewer possible points of conflict. You should probably start with pulling the latest code from upstream into your PR branch and resolve any merge conflicts due to changes that were merged since you submitted your PR.

> @akohlmey Does this fit into the planning for the patch release? Is there an approximate date considered for this patch?

This cannot be predicted at the moment. We have some plans for reorganizing packages and it depends on when we implement those. There is a good chance that the NN potentials will be added not to the upcoming patch release, but the one after that (tentatively in early May).

**Comment by singraber** [comment]
> No worries. Because of using an external library you PR is ahead of the curve, since there are fewer possible points of conflict. You should probably start with pulling the latest code from upstream into your PR branch and resolve any merge conflicts due to changes that were merged since you submitted your PR.
> 
Thank you for your suggestion, I will of course make sure that the PR is in no conflict with the master before requesting a review.
> 
> This cannot be predicted at the moment. We have some plans for reorganizing packages and it depends on when we implement those. There is a good chance that the NN potentials will be added not to the upcoming patch release, but the one after that (tentatively in early May).

Ok, thanks for the explanation and the perspective!

**Comment by akohlmey** [comment]
@singraber can you give us an update to the status of this PR?
If possible we would like to make this work with an automatically downloaded library like we have it with other packages, if the library is not already installed. Let us know, if you would need help with that.

**Comment by singraber** [comment]
I have made most of the requested changes except for the CMake suggestion to have an automated download. I will give it a try and adapt from `LATTE.cmake` but I may ask for some advice if I cannot make it work in the next days... to be honest, I find CMake very confusing and hard to learn.

Apart from that the only major change concerns the element <-> type mapping where I followed the example of `pair_style airebo` but I still have to update the docs accordingly.

**Comment by akohlmey** [comment]
Thanks for the status update. This helps us a lot organizing out work for the next weeks.

> I have made most of the requested changes except for the CMake suggestion to have an automated download. I will give it a try and adapt from `LATTE.cmake` but I may ask for some advice if I cannot make it work in the next days... to be honest, I find CMake very confusing and hard to learn.

True, but there is a very good documentation at https://cmake.org. It is quite a bigger challenge to learn CMake scripting to add to an existing package rather than starting from scratch. Please make sure that you consult the documentation for CMake version 3.10 or require a correspondingly newer CMake version if you have to use features that are only available in a newer version. Building LAMMPS is a rather complex procedure with many dependencies and challenging steps for any automated configuration/build system. There are not many alternatives and they would be even more challenging to use.

I am happy to give a helping hand with this, if you get stuck. I just need to know the URL from where the desired version of the library can be downloaded (by preference a specific version, so that we don't have to worry about incompatibilities and can do updates explicitly after testing for compatibility).

> Apart from that the only major change concerns the element <-> type mapping where I followed the example of `pair_style airebo` but I still have to update the docs accordingly.

Ok. that sounds good. Just let us know when you need help and/or when we should be taking a closer look again.

**Comment by akohlmey** [comment]
@singraber just wanted to check in on the status of the PR again. Can you let me know what is still missing from your side and how and where I could help to make it happen. I am currently looking over some nearly completed pull requests to see if they can be finished and merged before the pending patch release since we are going to make some changes after the patch release that will require additional changes to most pending pull requests.

**Comment by akohlmey** [comment]
> @akohlmey I am very sorry for these repeated delays from my side, I have to work in my free time on this which is currently very limited... thank you for your patience.

No problem. There are a few other PRs that require my attention, so you are not holding anything up currently.

> The good news is that it's almost completed! I have now adapted scripts for the traditional and CMake build process to allow for an automatic download, build and linking of n2p2 together with LAMMPS. Tomorrow I will bring the documentation up-to-date with the changes and do some final testing. Then the PR should be ready for another round of review.

Excellent!

**Comment by akohlmey** [comment]
@singraber I have started updating your PR branch to the latest upstream version and am looking into some "maintenance" changes. Would you prefer me to push those changes/updates directly into your PR branch or rather submit them as a pull request in your fork (I am asking since this is still flagged as draft and thus I am assuming you may have some more pending changes and I want to avoid disruptions or conflicts there. I have some unexpected opportunities to work on pending pull requests and this seemed the most "ready" target to get it updated first).  basically I am just trying to anticipate what would be done after to flagged this pull request as "ready for review".

**Comment by singraber** [comment]
@akohlmey You can of course directly push to my PR branch, thanks for asking! I am very close to making it ready for review, if nothing unexpected happens then I only need to change some parts of the documentation for the build process. In fact I will have a look at this right now...

**Comment by akohlmey** [comment]
> @akohlmey You can of course directly push to my PR branch, thanks for asking! I am very close to making it ready for review, if nothing unexpected happens then I only need to change some parts of the documentation for the build process. In fact I will have a look at this right now...

Thanks. Do you want me to push the update to upstream right now? Or should I wait until you are done (then I would resolve any additional merge conflicts with my changes).

There will be more changes from me following. I already noticed some issues with the build of the external library when it is downloaded by LAMMPS.

**Comment by singraber** [comment]
You can push it right now, I am working only on the `Build_extras.rst` and the `lib/hdnnp/README`, so there shouldn't be any major extra conflicts.

**Comment by akohlmey** [comment]
> You can push it right now, I am working only on the `Build_extras.rst` and the `lib/hdnnp/README`, so there shouldn't be any major extra conflicts.

Looks like you have that option disabled. So I'll need to go via the pull request route anyway.

**Comment by singraber** [comment]
Hmm.. the "Allow edits and access to secrets by maintainers " is checked tough!

**Comment by akohlmey** [comment]
> Hmm.. the "Allow edits and access to secrets by maintainers " is checked tough!

Then it could be the "draft" state of the PR that protects it from being pushed to. good to know. ;-)

**Comment by singraber** [comment]
Ok I made it ready for review.. maybe it works then.

**Comment by akohlmey** [comment]
> Ok I made it ready for review.. maybe it works then.

no such luck. same error:

```
$ git push
Enumerating objects: 538, done.
Counting objects: 100% (487/487), done.
Delta compression using up to 4 threads
Compressing objects: 100% (157/157), done.
Writing objects: 100% (349/349), 144.77 KiB | 72.38 MiB/s, done.
Total 349 (delta 278), reused 253 (delta 187), pack-reused 0
remote: Resolving deltas: 100% (278/278), completed with 102 local objects.
To github.com:lammps/lammps.git
 ! [remote rejected]       pair-style-nnp -> refs/pull/2626/head (deny updating a hidden ref)
error: failed to push some refs to 'git@github.com:lammps/lammps.git'

```

**Comment by singraber** [comment]
Sorry, then I don't know why it's not working.. I have never set anything special for my fork or this branch. I can see that I am an " outside collaborator" for lammps in my user settings, everything else is untouched as far as I can see.

However, I pushed the final changes to the docs just now.

**Comment by singraber** [comment]
Do you want me to pull in the latest changes from upstream LAMMPS now (this would usually be my last step) or would this mean to much interference with your work?

**Comment by akohlmey** [comment]
> Do you want me to pull in the latest changes from upstream LAMMPS now (this would usually be my last step) or would this mean to much interference with your work?

Thanks, but this is not needed. I already have your latest changes and resolved the merge conflicts. I'll just submit a pull request when I am done. This will take a bit, since I need to tinker with how the external library is being built and how the config settings from LAMMPS' CMake config are translated to your makefiles.

Some related questions.
1) did you ever try to compile for Windows or MacOS or other operating systems? 
2) I see you have MPI calls. How should it be handled if LAMMPS is compiled for serial use? exclude the package? use a make flag? use LAMMPS' dummy MPI library?

**Comment by singraber** [comment]
Ok great, a pull request is also fine.

Regarding your questions:

(1) I know that users can compile n2p2 on Windows and Mac but I have not tried it myself. Also, I could not test it for this PR because I don't have access to these operating systems.

(2) In case of a serial build, also n2p2 needs to be built without MPI (there is a special flag in n2p2's makefiles, see e.g. makefile.gnu, there is a N2P2_NO_MPI variable somewhere commented out by default.). That would then work together with a dummy library build of LAMMPS. However, I have not yet considered this for the automatic build options.. if you want I can have a look at this.

**Review by akohlmey — APPROVED** [review]

**Comment by akohlmey** [comment]
@sjplimp @athomps this is now ready for final review

**Review by sjplimp — APPROVED** [review]

**Comment by athomps** [comment]
I have a couple of minor changes that I would like to push shortly.

**Comment by akohlmey** [comment]
> I have a couple of minor changes that I would like to push shortly.

There have been difficulties pushing into this PR despite having the permission to push. I just had to resolve to some unusual tricks to resolve the merge conflict due to the URL changes.

I you have problems, please email me a `git diff` output and I'll see what I can do.

**Review by athomps — APPROVED** [review]
I approve this PR.

**Comment by athomps** [comment]
Done with edits. I only experienced the normal amount of difficulty :-).

**Comment by akohlmey** [comment]
> Done with edits. I only experienced the normal amount of difficulty :-).

You live a charmed life. Some higher entity must have recently decided to make me suffer for the sins I have committed in my life. :cry:

**Pull request merged**
